File: python-examples/subscription-modes/key_shared_subscription.py
#!/usr/bin/env python3
"""
Key_Shared 订阅模式示例
演示 Key_Shared 订阅模式的使用
"""

import pulsar
import threading
import time

SERVICE_URL = 'pulsar://localhost:6650'
TOPIC = 'persistent://public/default/key-shared-topic'

def consumer_worker(consumer_id):
    """Consumer 工作线程"""
    client = pulsar.Client(SERVICE_URL)
    
    try:
        consumer = client.subscribe(
            TOPIC,
            subscription_name='key-shared-subscription',
            consumer_type=pulsar.ConsumerType.KeyShared
        )
        
        print(f"   ✅ Consumer {consumer_id} 已启动")
        
        # 接收消息
        received_messages = []
        for _ in range(10):
            try:
                msg = consumer.receive(timeout_millis=3000)
                if msg:
                    content = msg.data().decode('utf-8')
                    key = msg.partition_key() if msg.partition_key() else "无Key"
                    print(f"   📩 Consumer {consumer_id} 接收: {content} (Key: {key})")
                    received_messages.append(content)
                    consumer.acknowledge(msg)
            except Exception:
                pass
        
        if received_messages:
            print(f"   ✅ Consumer {consumer_id} 共接收 {len(received_messages)} 条消息")
        
    except Exception as e:
        print(f"   ❌ Consumer {consumer_id} 错误: {e}")
    finally:
        client.close()

def main():
    client = pulsar.Client(SERVICE_URL)
    
    try:
        print("🚀 开始 Key_Shared 订阅模式示例...\n")
        
        # 创建 Producer
        producer = client.create_producer(TOPIC)
        
        # 发送带 Key 的消息
        print("📤 发送带 Key 的测试消息...")
        keys = ['user-1', 'user-2', 'user-3']
        for i in range(1, 13):
            key = keys[(i - 1) % 3]
            message = f"消息 {i}"
            producer.send(
                message.encode('utf-8'),
                partition_key=key
            )
            print(f"   ✅ 发送: {message} (Key: {key})")
        
        # 创建多个 Key_Shared Consumer
        print("\n📥 创建 3 个 Key_Shared Consumers...")
        print("   ℹ️  相同 Key 的消息会被同一个 Consumer 接收\n")
        
        threads = []
        for i in range(1, 4):
            thread = threading.Thread(target=consumer_worker, args=(i,))
            thread.start()
            threads.append(thread)
        
        # 等待所有线程完成
        for thread in threads:
            thread.join()
        
        print("\n✅ Key_Shared 订阅模式示例执行完成")
        print("💡 特点: 相同 Key 的消息保证顺序，不同 Key 可并行消费")
        
    except Exception as e:
        print(f"❌ 执行失败: {e}")
    finally:
        client.close()

if __name__ == '__main__':
    main()

File: go-examples/subscription-modes/key_shared_subscription.go
package main

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/apache/pulsar-client-go/pulsar"
)

func main() {
	client, err := pulsar.NewClient(pulsar.ClientOptions{
		URL: "pulsar://localhost:6650",
	})
	if err != nil {
		log.Fatal(err)
	}
	defer client.Close()

	topic := "persistent://public/default/go-key-shared-topic"
	fmt.Println("🚀 开始 Key_Shared 订阅模式示例...\n")

	// 创建 Producer 并发送带 Key 的消息
	producer, err := client.CreateProducer(pulsar.ProducerOptions{Topic: topic})
	if err != nil {
		log.Fatal(err)
	}

	keys := []string{"order-001", "order-002", "order-003"}
	fmt.Println("📤 发送带 Key 的消息:")
	for round := 1; round <= 3; round++ {
		for _, key := range keys {
			msg := fmt.Sprintf("%s-事件-%d", key, round)
			_, err := producer.Send(context.Background(), &pulsar.ProducerMessage{
				Key:     key,
				Payload: []byte(msg),
			})
			if err != nil {
				log.Fatal(err)
			}
			fmt.Printf("   ✅ 发送: Key=%s, Payload=%s\n", key, msg)
		}
	}
	producer.Close()

	// 创建两个 Key_Shared Consumer
	fmt.Println("\n📥 创建 Key_Shared Consumer...")
	var wg sync.WaitGroup

	for idx := 1; idx <= 2; idx++ {
		consumer, err := client.Subscribe(pulsar.ConsumerOptions{
			Topic:            topic,
			SubscriptionName: "go-key-shared-sub",
			Type:             pulsar.KeyShared,
			Name:             fmt.Sprintf("consumer-%d", idx),
		})
		if err != nil {
			log.Fatal(err)
		}
		fmt.Printf("   ✅ Consumer-%d 已创建\n", idx)

		wg.Add(1)
		go func(c pulsar.Consumer, id int) {
			defer wg.Done()
			defer c.Close()
			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()
			for {
				msg, err := c.Receive(ctx)
				if err != nil {
					return
				}
				fmt.Printf("   📩 [Consumer-%d] Key=%s, Payload=%s\n",
					id, msg.Key(), string(msg.Payload()))
				c.Ack(msg)
			}
		}(consumer, idx)
	}

	fmt.Println("   ℹ️  相同 Key 的消息始终路由到同一个 Consumer\n")
	wg.Wait()

	fmt.Println("\n✅ Key_Shared 订阅模式示例执行完成")
	fmt.Println("💡 特点: 按 Key 分组顺序消费，兼顾并行度和顺序性")
}
